### PR TITLE
implement CachedFrozenArray

### LINKED ISSUE #34049: Create CachedFrozenArray abstraction
There are a number of pieces of code that have the following patterns:
* a member field like `DomRefCell<Option<Heap<JSVal>>>`
* an implementation that first checks if there is a cached value present in the member field and returns it
* if there is no cached value present, compute a list of values, call to_frozen_array, and store the result in the member field, then return the value

There's always a comment about performing the operation in a particular order to maintain the safety invariants of the JS engine integration. This makes it a very good candidate for abstracting it into a type that each of these pieces of code can use instead! I'm imagining that the DomRefCell field moves into the new CachedFrozenArray type, and that type has one interface:
```rust
fn get_or_init<F: FnOnce() -> Vec<T>, T: ToJSValConvertible>(&self, f: F) -> JSVal;
```
It will accept a closure very similar to [MutNullableDom::or_init](https://doc.servo.org/script/dom/bindings/root/struct.MutNullableDom.html#method.or_init), allowing callers to provide the unique initialization logic while avoiding repeating the caching and JS engine safety logic.

* https://github.com/servo/servo/blob/a501572971d7ec6761aeab1f8f5dba00158bc1a6/components/script/dom/messageevent.rs#L305
* https://github.com/servo/servo/blob/a501572971d7ec6761aeab1f8f5dba00158bc1a6/components/script/dom/extendablemessageevent.rs#L210
* https://github.com/servo/servo/blob/a501572971d7ec6761aeab1f8f5dba00158bc1a6/components/script/dom/globalscope.rs#L3074

### PR DESCRIPTION
**Pull request by Gae24**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #34049  (GitHub issue number if applicable)

- [X] These changes do not require tests because no new funtionality

### CODE REVIEW

**Review by jdm — APPROVED** [review]
Thank you! This is a lovely cleanup.

**Comment by Gae24** [comment]
> Thank you! This is a lovely cleanup.

Well not so much...

**Comment by Gae24** [comment]
When rebasing after #34087, I didn't even think that was borrowing two times.

File: components/script/dom/bindings/frozenarray.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    pub fn get_or_init<F: FnOnce() -> Vec<T>, T: ToJSValConvertible>(
+        &self,
+        f: F,
+        cx: JSContext,
+        mut retval: MutableHandleValue,
+    ) {
+        if let Some(inner) = &*self.frozen_value.borrow() {
+            retval.set(inner.get());
+        }
```
Comment (jdm) [inline components/script/dom/bindings/frozenarray.rs:34]: I think we need to return here instead of continuing.

**Pull request merged**